### PR TITLE
Activate P2MR validation weight v2 on testnet4 at height 60,000

### LINKED ISSUE #102: Prevent `mr(multi_a)` from producing annexless satisfactions that exhaust validation weight
## Summary

The built-in P2MR descriptor and signing path can create high-threshold, single-leaf `mr(multi_a(...))` outputs whose normal annexless satisfaction is rejected with `SCRIPT_ERR_P2MR_VALIDATION_WEIGHT`.

The smallest affected case is a direct 5-of-5 leaf with default-sighash 3,680-byte signatures. The descriptor accepts the output, but the annexless witness has a validation budget of 18,641 and five successful PQC checks cost 18,650, leaving a deficit of 9.

This is a real wallet/descriptor safety problem, but it is narrower than permanent consensus-level fund loss. Extra committed script bytes, a deeper control path, or a consensus-valid annex increase the budget. Annexes are currently nonstandard and the built-in signing code only supports annexless signatures, so those alternatives do not make the default output safely spendable through the normal wallet and relay path.

## Verified on

`origin/1.0.0` at `988756471aeecdf4463c04be49da2b7b89a98c21`.

## Source trace

- Each non-empty P2MR signature operation costs 3,730 units: [`src/script/script.h#L70-L77`](https://github.com/Qbit-Org/qbit/blob/988756471aeecdf4463c04be49da2b7b89a98c21/src/script/script.h#L70-L77), [`src/script/interpreter.cpp#L403-L417`](https://github.com/Qbit-Org/qbit/blob/988756471aeecdf4463c04be49da2b7b89a98c21/src/script/interpreter.cpp#L403-L417).
- The budget is the serialized size of the entire input witness plus 50: [`src/script/interpreter.cpp#L2389-L2399`](https://github.com/Qbit-Org/qbit/blob/988756471aeecdf4463c04be49da2b7b89a98c21/src/script/interpreter.cpp#L2389-L2399).
- PQC signatures are 3,680 bytes: [`src/crypto/pqc.h#L20-L27`](https://github.com/Qbit-Org/qbit/blob/988756471aeecdf4463c04be49da2b7b89a98c21/src/crypto/pqc.h#L20-L27).
- The P2MR descriptor builds 34 bytes per key plus the threshold and `OP_NUMEQUAL`, and does not reject unsafe threshold/key-count combinations: [`src/script/descriptor.cpp#L1729-L1804`](https://github.com/Qbit-Org/qbit/blob/988756471aeecdf4463c04be49da2b7b89a98c21/src/script/descriptor.cpp#L1729-L1804).
- P2MR signing explicitly constructs annexless signature hashes: [`src/script/sign.cpp#L108-L137`](https://github.com/Qbit-Org/qbit/blob/988756471aeecdf4463c04be49da2b7b89a98c21/src/script/sign.cpp#L108-L137).
- P2MR annexes are consensus-valid and included in the original witness size used for the budget, but policy rejects them as nonstandard: [`src/script/interpreter.cpp#L2345-L2399`](https://github.com/Qbit-Org/qbit/blob/988756471aeecdf4463c04be49da2b7b89a98c21/src/script/interpreter.cpp#L2345-L2399), [`src/policy/policy.cpp#L354-L383`](https://github.com/Qbit-Org/qbit/blob/988756471aeecdf4463c04be49da2b7b89a98c21/src/policy/policy.cpp#L354-L383).

## Exact 5-of-5 calculation

For a direct, depth-0 `mr(multi_a(5,...))` using raw default-sighash signatures:

| Witness component | Serialized bytes |
| --- | ---: |
| Witness item count | 1 |
| Five signatures: `5 * (CompactSize(3680) + 3680)` | 18,415 |
| 172-byte leaf script plus length prefix | 173 |
| One-byte control block plus length prefix | 2 |
| Validation offset | 50 |
| **Initial budget** | **18,641** |
| Five checks at 3,730 each | **18,650** |
| **Margin** | **-9** |

Non-default-sighash signatures add one byte each but still leave the 5-of-5 case short by 4.

For default-sighash, direct single-leaf failures begin as follows:

| Descriptor | Margin |
| --- | ---: |
| 5-of-5 | -9 |
| 6-of-6 | -22 |
| 7-of-7 | -35 |
| 8-of-8 | -46 |
| 8-of-9 | -11 |
| 9-of-9 | -59 |

By contrast, 2-of-3, 3-of-5, and 5-of-7 have positive margins.

## Important scope correction

This is not true for every 5-of-5 P2MR script and is not an absolute consensus lock:

- A control path one level deeper adds 32 serialized bytes and makes the direct 5-of-5 budget positive.
- Additional committed, executable-safe script bytes also increase the budget.
- An 8-byte annex adds 9 serialized bytes and makes the depth-0 default-sighash 5-of-5 witness exactly meet the budget.

However, annex spending is nonstandard and unsupported by the normal signer, and these workarounds must have been anticipated when the output or signatures were constructed. The built-in direct descriptor remains unsafe by default.

## Expected behavior

Every descriptor configuration accepted for funding should have a standard, wallet-constructible satisfaction that meets consensus validation-weight limits.

## Fix options

Because mainnet has not launched, prefer aligning the consensus validation-weight accounting so standard descriptor satisfactions do not require padding. Options include:

- reduce `VALIDATION_WEIGHT_PER_SIGOP_PQC` to match the actual marginal witness budget while preserving the intended block-wide verification bound; or
- add deterministic per-signature headroom to the initial P2MR budget.

If consensus accounting is intentionally left unchanged, reject unsafe `mr(multi_a)` configurations during descriptor/address creation and PSBT construction, with a clear error explaining the threshold and budget. This avoids new unsafe outputs but does not provide a normal recovery path for outputs already created.

## Regression coverage

- Add a table-driven test for direct `k`-of-`n` P2MR descriptor satisfactions around every budget boundary.
- Verify that every descriptor accepted for funding can be finalized and passes `VerifyScript` under mandatory and standard flags.
- Include raw/default and one-byte non-default sighash signatures.
- Include depth-0 and nonzero-depth control blocks so the test documents that control-path size affects the budget.
- If annex padding remains a recovery mechanism, test it separately as consensus-valid but nonstandard behavior.

## Acceptance criteria

- The wallet/descriptor layer cannot silently create an output whose normal satisfaction is rejected for P2MR validation weight.
- The 5-of-5 boundary has explicit regression coverage.
- The chosen accounting preserves a documented upper bound on PQC verification work.
- Error messages identify unsupported threshold/key-count combinations if creation is blocked.

### PR DESCRIPTION
**Pull request by kiwidream**
## Summary

Closes #102.

P2MR validation-weight v2 charges each successful PQC signature operation by
its marginal serialized witness size: 3,683 units for a 3,680-byte signature
and its three-byte CompactSize prefix. The previous 3,730-unit debit combined
the signature size with the validation offset, causing normal annexless
`mr(multi_a(...))` satisfactions such as direct 5-of-5 leaves to fail despite
being produced by the descriptor and wallet paths.

Mainnet uses the corrected accounting from genesis. Testnet3, signet, and
default regtest do as well. Testnet4 retains the legacy debit below block
60,000 and activates v2 starting with block 60,000. This is a coordinated
consensus relaxation, not a testnet reset: genesis, network identity, and all
pre-activation history remain unchanged.

This change also:

- applies candidate-height rules consistently in mempool and block validation;
- revalidates mempool scripts when a reorg crosses the activation boundary;
- reports the applicable rule through `getblockchaininfo`, `getblocktemplate`,
  and `createauxblock`;
- prevents the signer from consuming stateful PQC signatures for infeasible
  P2MR plans;
- identifies the 35-signature P2MR v1 initial-stack boundary in descriptor
  tooling, rejecting unsafe active imports and normal address derivation while
  permitting inactive recovery imports with a warning; and
- documents activation, operator impact, and the accounting rationale.

## Testing

- [x] Built locally.
- [x] Ran focused unit or functional tests for the changed area.
- [x] Ran lint or formatting checks relevant to this change.
- [ ] Not run. Reason:

Checks performed:

- CMake native build with GUI, benchmarks, and fuzz binary disabled.
- 140 focused descriptor, P2MR script, and transaction unit tests.
- 11 mempool, validation, and chainstate regression tests.
- P2MR validation-weight activation functional test, including forward
  activation, mining, reverse-boundary reorg eviction, and reconsideration.
- Descriptor RPC functional test.
- RPC documentation tests.
- Explicit testnet4 activation-parameter unit test.
- Full lint suite using the repository Docker lint image.
- `git diff --check`.

## Target Branch

- [x] This PR targets `main` or a maintainer-requested release branch such as `0.1.x`.

Target: `1.0.0`.

## Risk / Review Notes

- [x] Consensus, script, crypto, wallet, P2P, release, CI, or security-sensitive behavior changed.
- [ ] No consensus, script, crypto, wallet, P2P, release, CI, or security-sensitive behavior changed.

Notes:

Testnet4 miners and validating nodes must upgrade before height 60,000. Older
nodes reject blocks that exercise the newly available validation budget and
can remain on a stale fork. The activation changes no previously valid block.

## Docs / Process Impact

Choose exactly one:

- [x] I updated public docs because this PR changes user-visible behavior, integration guidance, release/process guidance, or expected validation.
- [ ] No public docs update needed. Reason:

## libbitcoinpqc Subtree Checklist (if `src/libbitcoinpqc` changed)

Not applicable; `src/libbitcoinpqc` is unchanged.

- [ ] Source commit is reachable from an immutable release tag in `Qbit-Org/qbit-libbitcoinpqc`.
- [ ] qbit imports the tagged upstream tree directly without pruning or a curated subtree branch.
- [ ] Subtree import/update was performed with `contrib/devtools/update-libbitcoinpqc-subtree.sh`.
- [ ] `test/lint/libbitcoinpqc-subtree-check.sh` passes locally.
- [ ] Any default tag change in `contrib/devtools/update-libbitcoinpqc-subtree.sh` is intentional and matches `doc/subtrees/libbitcoinpqc.md`.

---
<a href="https://app.blacksmith.sh/Qbit-Org/codesmith/qbit/pr/109"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1786388334&installation_id=141183937&pr_number=109&repository=Qbit-Org%2Fqbit&return_to=https%3A%2F%2Fgithub.com%2FQbit-Org%2Fqbit%2Fpull%2F109&signature=eaf2e047816575902251cd76f7839ff583243170a8099d7597ae9c048c74cb72"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

---

> [!NOTE]
> **High Risk**
> Consensus script validation and testnet4 height-activated accounting change which blocks and mempool txs are valid; operators must upgrade before testnet4 block 60,000 or risk stale forks.
> 
> **Overview**
> Introduces **P2MR validation-weight v2**, lowering the per successful PQC signature debit from **3,730** to **3,683** units (serialized signature + CompactSize). **Mainnet and most networks use v2 from genesis**; **testnet4** keeps the legacy debit until **block 60,000** without a chain reset—a coordinated consensus relaxation.
> 
> Consensus and validation now pick legacy vs v2 via `SCRIPT_VERIFY_P2MR_LEGACY_VALIDATION_WEIGHT` and height-gated params; **mempool and block checks use the next-block/candidate height**, and **reorgs re-run script validation** so txs that only fit under v2 drop off when rolling back past activation.
> 
> **RPC** exposes `p2mr_validation_weight` on `getblockchaininfo`, `getblocktemplate`, and `createauxblock`. **Wallet/descriptor** paths reject or flag P2MR trees that cannot satisfy the **35-signature** initial-stack cap (`HasP2MRStandardSatisfaction`, signing plan limits). Docs, release notes, regtest `-testactivationheight=p2mrweightv2@`, and functional tests cover activation and 5-of-5 spends.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 161cde6447155e11530f7f645c503441b98d5013. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/script/descriptor.h [hunk 1]
```diff
@@ -127,6 +127,10 @@ struct Descriptor {
      *  This is true for all descriptors except ones that use `raw` or `addr` constructions. */
     virtual bool IsSolvable() const = 0;
 
+    /** Whether a P2MR descriptor has at least one wallet-constructible
+     * satisfaction within the v1 initial-stack resource limits. */
+    virtual bool HasP2MRStandardSatisfaction() const { return true; }
```
[the remark this answers is not in the document]
Comment (kiwidream) [reply src/script/descriptor.h]: Addressed in 815ad414a3. `HasP2MRStandardSatisfaction()` now defaults to `IsSolvable()`, so unsolvable P2MR `addr()` and `raw()` descriptors report `false` while the specialized P2MR threshold checks remain intact. Added `getdescriptorinfo` regression coverage for `rawmr()`, `addr()`, and `raw()` P2MR forms. Verified with the descriptor P2MR unit suite and `rpc_getdescriptorinfo.py`.

File: src/script/sign.cpp [hunk 2]
```diff
@@ -561,8 +562,19 @@ static bool BuildP2MRScriptSigningPlan(const SigningProvider& provider, const Ba
...
+        if (item.size() > MAX_P2MR_V1_STACK_ITEM_SIZE) return false;
+        initial_stack_bytes += item.size();
+        if (initial_stack_bytes > MAX_P2MR_V1_TOTAL_INITIAL_STACK_BYTES) return false;
+    }
     estimated_stack.emplace_back(script_bytes.begin(), script_bytes.end());
     estimated_stack.push_back(control_block);
+    if (GetSerializeSize(estimated_stack) + VALIDATION_WEIGHT_OFFSET <
+        static_cast<size_t>(num_estimated_signatures) * P2MR_VALIDATION_WEIGHT_PER_SIGOP_V2) {
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (kiwidream) [reply src/script/sign.cpp:577]: The partial-plan check is intentional under v2. Each 3,680-byte PQC signature contributes 3,683 serialized witness bytes including its CompactSize prefix, exactly matching `P2MR_VALIDATION_WEIGHT_PER_SIGOP_V2`. Adding another cosigner signature therefore increases the validation budget by at least its full debit, so a partial plan cannot become underfunded solely by reaching the threshold. Explicit-hashtype signatures contribute one additional byte. Threshold, item-size, and aggregate-stack limits are checked separately. The incremental 2-of-3 PSBT unit test and independent-wallet 5-of-5 activation functional test cover the multiparty flow. Legacy accounting is intentionally different and is tested at the activation boundary.

**Comment by kiwidream** [comment]
## Findings

1. **[P2] Preserve legacy policy before delayed P2MR activation**

   [`GetStandardScriptFlagsForNextBlock()`](https://github.com/Qbit-Org/qbit/blob/fd0790b752bb5c69c69b663f226800d51c7e1b99/src/validation.cpp#L2444-L2455) adds the legacy debit only when `next_height >= P2MRHeight`, even though `STANDARD_SCRIPT_VERIFY_FLAGS` already enables P2MR policy before consensus activation.

   With `p2mr@H` and `p2mrweightv2@H+K`, a v2-only 5-of-5 transaction can enter the mempool at tip `H-2`. If block `H-1` arrives without it, candidate block `H` uses legacy rules, but a normal forward connection does not run the new reorg filter. The now-invalid transaction remains available to `BlockAssembler`, where it can make `TestBlockValidity` fail.

   Select the policy debit for the next height where P2MR rules apply—for example, using `max(next_height, P2MRHeight)`—or revalidate the mempool on the forward P2MR boundary. Please add a regression combining delayed P2MR and delayed validation-weight-v2 activation.

2. **[P2] Document the block-wide PQC verification bound**

   The [new reference](https://github.com/Qbit-Org/qbit/blob/fd0790b752bb5c69c69b663f226800d51c7e1b99/doc/reference/p2mr-validation-weight-v2.md#L3-L12) explains the per-input budget but never derives the block-wide maximum or relates it to the repository’s PQC verification target. This leaves an explicit acceptance criterion from #102 unmet for a consensus relaxation.

   Please derive a numeric upper bound from `MAX_BLOCK_WEIGHT`, per-input offsets, and minimum P2MR serialization; compare legacy and v2 maximum successful checks; and translate the delta into validation time using the maintained PQC benchmark.

3. **[P3] Add the requested boundary regression matrix**

   The new coverage verifies direct default-sighash 5-of-5 and reports the 35/36 descriptor boundary, but it does not add the issue’s requested table of `k`-of-`n` boundaries, non-default sighashes, and nonzero control depths. It also does not demonstrate finalization plus mandatory/standard `VerifyScript` behavior across those accepted boundaries.

   Please make the accounting invariant executable with table-driven coverage, especially 5-of-5 and the 35/36 initial-stack boundary under default and explicit sighashes.

## Consensus Impact

Consensus-critical. This is a height-activated testnet4 consensus relaxation; old nodes can stale-fork when a block first exercises v2-only budget. I found no production-parameter mismatch in block-height selection. Finding 1 concerns supported regtest activation combinations and mining-template correctness.

## Review Scope

Reviewed PR #109 at `fd0790b752bb5c69c69b663f226800d51c7e1b99`, backing issue #102, all 28 changed files, prior review threads, mempool/block/reorg call paths, wallet signing and descriptor paths, RPC candidate-height reporting, documentation, and CI status.

## Deep Checks

- Confirmed block validation selects legacy/v2 from the block’s own height.
- Confirmed next-block mempool checks and reverse-boundary revalidation use flag-distinct script-cache entries.
- Checked transaction- and data-signature opcodes use the same debit.
- Disproved the partial-cosigner concern: under v2, the final canonical witness retains positive margin from leaf/control serialization, empty elements, and the validation offset.
- Confirmed 35 explicit-hashtype signatures consume 128,835 initial-stack bytes, while 36 default signatures exceed 128 KiB.
- Confirmed unsolvable `rawmr`, `addr`, and `raw` descriptors now report no standard satisfaction.

## Issue Fit

The PR fixes the direct 5-of-5 accounting failure, adds candidate-height activation behavior, protects stateful signing from structurally infeasible plans, exposes activation state, and provides explicit 5-of-5 coverage. The documented work bound and requested regression matrix remain incomplete.

## qbit-Specific Checks

Reviewed stateful PQC counter ordering, P2MR initial-stack limits, signature serialization, testnet4 hard-fork coordination, reverse-reorg eviction, AuxPoW template reporting, and no-witness-discount block accounting.

## Validation Reviewed

- All reported GitHub checks are green at the reviewed head, including full validation, sanitizers, platform builds, focused units, functional smoke, RPC/public-doc checks, and Docker lint.
- `git diff --check origin/1.0.0...origin/pr-109` passes.
- No additional local compilation was performed.

## Residual Risk

The PR merge-base predates three current `1.0.0` commits, including another `validation.cpp` change. Git reports it mergeable, but the combined result should be retested after updating the branch. Live testnet4 height and operator upgrade readiness were not independently verified.
